Add 'max_sub_interval' option to derivative sensor

karwosts opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
Copying the implementation from #110685, add a `max_sub_interval` option to derivative sensor, which will force the derivative recalculation at the requested time interval if no updates come from the source sensor during that time.

Enabling this option will allow the derivative sensor to return to zero when the source stops changing. Presently the derivative will just forever keep the last non-zero value, which is not mathematically expected.

E.g.:
![image](https://github.com/user-attachments/assets/5c982c36-6db1-47eb-9e2f-f35577561f48)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #83496
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39690

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

karwosts committed:
Add 'max_sub_interval' option to derivative sensor

karwosts committed:
add strings

karwosts committed:
little coverage

karwosts committed:
improve test accuracy

karwosts commented:
Not quite sure what's wrong with codecov/project, but I don't believe it's related to this change, it just broke when I merged dev to fix a merge conflict. Not sure if there's a way to revive it or if it just needs to be ignored.

mark007 commented:
Hoping this one can be reviewed / merged. It affects many of us, especially in the likes of our house where many sensors update very slowly/infrequently, eg where temperatures drop/rise very slowly. All I can ask is that those with the power to do so, review/merge it sooner rather than later, thanks!

gjohansson-ST commented:
Would this be solved by implementing also tracking reported and not only changed values?
I assume it's not expected that the input sensors stops reporting?
I guess it's not exactly the same as this PR but it would be less complicated perhaps to solve the issue.

mekaneck commented:
I think using the `state_reported` event to update the derivative is a much needed change, but it should be a separate PR. Both changes are necessary to solve the problem correctly and completely. Both changes were implemented on the Riemann Sum integration (2 separate PR's) and I don't see a reason anything should be different for the derivative integration.

Anton2079 commented:
I also want to nudge this as I’m also experiencing this issue.

Anton2079 commented:
Sorry to ping when you're not listed, however @ThomDietrich are you able to review this, it is against @afaucogney but he hasn't been seen on Github since August.
This issue has been observed since December 2022.
https://github.com/home-assistant/core/issues/83496

ThomDietrich commented:
Hey guys.
The discussed issue is very similar to a problem that we have over in the statistics component. See [here](https://github.com/home-assistant/core/pull/127305#issuecomment-2428873449) and [here](https://github.com/home-assistant/core/pull/127305#issuecomment-2431335581). Maybe that's why @Anton2079 mentioned me here.

I welcome the fix here. As a user of the derivative component myself, I run into this problem too.

1. I agree with @gjohansson-ST, the component should listen to the newly introduced `async_track_state_report_event`. I also agree with @karwosts, this should be a separate PR. However, I think the state report event PR would be the better and more significant improvement.
2. I agree with the idea to add intermediary updates.
    a. @karwosts did you consider a service based approach instead? I am not sure if a fixed "sub interval" will solve this cleanly. A separate automation triggering an intermediary update would allow more control to the user, e.g. triggering an update just-in-time when another action needs the derivative value.
    b. The feature of a fixed sub interval is still a good fallback to the said service.

baudneo commented:
Patiently waiting for this to make its way to release. Really glad there's work being done to fix the derivative sensor.

sophof commented:
@karwosts Hi, I've started on reviewing your PR after talking to you on discord, but I have some generic questions first before I go ahead and write up my suggested changes and such. To be honest, I think there will be few that will take some work..

I think it is clear that extra codeowners are needed for this integration based on the time this has spent in limbo currently. Besides that, according to the architecture guidelines, when adding a feature (such as here) one should also add themselves as codeowner. Are you willing to do that? You also implied that you didn't really know much anymore because of the time that had passed since submitting the PR. I interpreted from that, that you might not be that motivated.

As an alternative I could submit my PR and add myself as codeowner. It has similar ideas, but refactors the code even more (I might remove some of it to put it in a later PR to make things simpler). The only work that is left for me is to copy your unit tests and naming scheme, which I like better than mine. Then you could review my code.

So my question boils down to: are you willing to be code-owner? And if not, are you willing to review my PR to help it along instead?

elupus commented:
Looks like reported event will be seen as opt in by integrations, so we must support something like this and probably enabled by default.

karwosts commented:
I'm thinking this has reached a dead-end. There was perhaps too much refactoring included in the initial PR, makes it probably too hard to review, and worse it is too hard to keep up to date with other changes.

The recent merging of state_reported now will require a significant amount of work to rewrite this again, and I'm not sure if it would do any good to attempt it.

Could try to split the necessary refactoring out into a separate PR before attempting any functional changes, but given that this integration has been abandoned by the codeowner I have no reason to think that would ever be reviewed either.

@sophof - if you have a version of this ready to go that has incorporated the state_reported change, I'm happy to close this and let you take it over if you would like to try to do so. 

If not, I think this needs some direction from a maintainer on where to go from here, before I'm willing to invest any more time on it.

karwosts commented:
I think the thinking is that it fixes it for some cases, but not all. Getting state_reported events from a constant sensor can't be guaranteed.

NoRi2909 commented on homeassistant/components/derivative/strings.json:
```diff
@@ -11,12 +11,14 @@
           "time_window": "If set, the sensor's value is a time weighted moving average of derivatives within this window.",
-          "unit_prefix": "The output will be scaled according to the selected metric prefix and time unit of the derivative."
+          "unit_prefix": "The output will be scaled according to the selected metric prefix and time unit of the derivative.",
+          "max_sub_interval": "Recalculates derivative if the source did not change for this duration. Use 0 for no time based updates."
```
Adding a hyphen to "time-based" should improve readability and (machine) translations.

emontnemery commented:
@karwosts do you mean you've given up on this PR? Be as it may, consuming `state_reported` is the "correct" solution, although it will obviously be a long time before integrations have removed their own suppressing of equal states, so the `max_sub_interval` adds value for sure.

karwosts commented:
> @karwosts do you mean you've given up on this PR?

Not entirely sure myself. If a member confirmed that they want to take this change, and that the initial approach was a reasonable approach that could be approved, I would consider taking another go at rewriting it to merge again. But the code owner seems not around anymore, so not really getting much feedback if it's gonna be considered.

karwosts commented:
> OK. Would you be interested in stepping up as code owner for this integration?

Sure I suppose I could give it a shot. I'm still pretty unfamiliar with a lot of the core concepts 😅 , but I guess this would be a quiet corner to learn. 

Perhaps I'll try one more PR with just the refactoring needed, and then when that is done can come add this new option on top of that.

karwosts committed:
reimplement at dev head

karwosts committed:
string

karwosts committed:
handle unavailable

karwosts committed:
simplify

karwosts committed:
Add self to codeowner

karwosts committed:
fix on remove

karwosts commented:
- Rebased and fixed conflicts (due to prior refactoring by other contributors this is actually a simpler now). 
- Added self to codeowners as requested. 

Ready for new review.

emontnemery commented on homeassistant/components/derivative/sensor.py:
```diff
@@ -192,6 +213,38 @@ def __init__(
+    def _calc_derivative_from_state_list(self, current_time: datetime) -> Decimal:
+        def calculate_weight(start: datetime, end: datetime, now: datetime) -> float:
+            window_start = now - timedelta(seconds=self._time_window)
+            if start < window_start:
+                weight = (end - window_start).total_seconds() / self._time_window
+            else:
+                weight = (end - start).total_seconds() / self._time_window
+            return weight
```
I'd suggest to remove the branching here to avoid duplicating the calculations
```suggestion
            return (end - max(start, window_start).total_seconds() / self._time_window
```

emontnemery commented on homeassistant/components/derivative/strings.json:
Please keep the strings sorted alphabetically

karwosts replied:
Sorted.

emontnemery reviewed (changes requested):
Looks good, just minor remarks.
Also, I don't see a link to updated user documentation.

karwosts committed:
Update homeassistant/components/derivative/sensor.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

karwosts committed:
Fix parenthesis

karwosts committed:
sort strings

emontnemery reviewed (approved):
Thanks, @karwosts 👍

baudneo commented:
Hell yeah

Anton2079 commented:
Omg omg omg
Can’t wait to test it
